### DISCUSSION TITLE
High-Level Use Cases

### DISCUSSION THREAD

**Discussion by thanethomson** [informalsystems/tendermint-rs#885] [General]
Some time ago there was discussion around capturing use cases for `tendermint-rs`. Let's collect these use cases here.

The questions we want to answer:
1. What Tendermint-related things do people want to do in Rust?
2. What of (1) should be included in the `tendermint-rs` repository and what should be considered a separate repository?

At present, we'd ideally like to move towards making `tendermint-rs` a **framework** for doing useful Tendermint-related things. Binaries would naturally then be moved to/created in separate repositories, as we can then separate framework-related issues from operational ones.

Feel free to comment below if you'd like your use case added!

## Current Use Cases

These are uses cases where we know the people making use of current released code and why they're using it.

* [Cosmos] needs core Tendermint data structures for its Rust SDK: https://github.com/cosmos/cosmos-rust
* [Informal Systems] is implementing IBC in Rust, which requires Tendermint compatibility: https://github.com/informalsystems/ibc-rs
* [Iqlusion] implements the Tendermint KMS, which requires Tendermint compatibility: https://github.com/iqlusioninc/tmkms

## Envisaged Use Cases

These are use cases where we know the people and what they want to do in relation to Tendermint in Rust, but the code's either a WIP or does not yet exist.

* [Informal Systems] wants to implement the following types of nodes, compatible with [Tendermint](https://github.com/tendermint/tendermint/):
  * Seed node
  * Sentry node
  * Full node
  * Validator

[Cosmos]: https://cosmos.network/
[Informal Systems]: https://informal.systems/
[Iqlusion]: https://www.iqlusion.io/

**Comment by tony-iqlusion** [comment]
At a high level I'd say @iqlusioninc's use cases boil down to the following two coarse grained areas:

## Core data structures / types

In libraries and apps where we want to consume chain state / "metadata", we would like a common set of rich types for representing as much.

Some concrete examples which weren't mentioned above:

- [`mintscan.rs`](https://github.com/iqlusioninc/crates/tree/main/mintscan) - API client for the Mintscan Explorer, which we use for black box observability of validator state / missed blocks
- [`sagan`](https://github.com/iqlusioninc/sagan) - multichain observability tool for Tendermint blockchains

Notably these use cases do not require the cryptography libraries which are currently a requirement of tendermint-rs.

Another example that might surprisingly fit this use case is [`delphi`](https://github.com/iqlusioninc/delphi), our (currently Terra-only) oracle service. We've designed it such that it only talks to edge network services and proposes oracle transactions to be signed, which are actually signed by TMKMS.

## Cryptography / signing

As previously noted, TMKMS and cosmos-rust provide validator and transaction signing capabilities respectively.

Separating the "core data structures" versus "cryptography" use cases has been proposed before, and was something I brought up again in #892, where I suggested spinning out a `tendermint-crypto` crate. Right now tendermint-rs has a number of hard cryptographic dependencies that aren't relevant to the first use case (and full disclosure, I'm the author/maintainer of several of those, so I'm suggesting something that would hurt my own download numbers for those dependencies, but make my life easier in the cases where we're just consuming chain data)
